Full codebase review

ekiefl opened the pull request:
# What is this

This codebase has matured without any code review. To right our wrongs, this PR introduces the entire codebase as additions so every line can be reviewed.

This was accomplished with two branches, `ek/empty-repo` and `ek/full-codebase-review`. `ek/empty-repo` branches off main and deletes the codebase (41b73b812d8bbcf42a690859ab1953f3481663f9). `ek/full-codebase-review` branches off `ek/empty-repo` and restores the codebase by committing the results of `git restore --source=e8c513dd4e2df0761d930591eae24575e4a2c671 .` as a single commit (919d56421971fdcb623d846d190f33bfbfa584df).

This PR requests a merge of `ek/full-codebase-review` into `ek/empty-repo`. Once merged, `ek/empty-repo` (which will by then be *not* empty), will be merged into `main` (without review). This control flow allows `main` to exist in its current state while improvements are made in this PR.

# Overview of codebase

This is a template repo for notebook publications. Scientists clone this repo as a template to construct their own notebook pubs. The template also serves as a demo, which is hosted here: https://arcadia-science.github.io/notebook-pub-template/

Briefly, the publication is rendered using Quarto. The publication is hosted by GitHub Pages through a GitHub Action that triggers when `main` is merged into `publish`. Rationale for this control flow provided in `README.md`.

Description of codebase:

* `_extensions`: These are [Quarto Extensions](https://quarto.org/docs/extensions/). Quarto extensions work by directly vendoring the extension source code, hence why this directory exists.

* `_freeze`: Quarto provides instructions for hosting via GitHub pages. This repo uses the [computation freezing](https://quarto.org/docs/publishing/github-pages.html#freezing-computations) instructions, where the cell outputs are executed locally and directly committed to the repo. This offloads the responsibility of directly running the notebook computations on our CI. Instead, our CI needs only to find the relevant cell outputs in `_freeze` when rendering the publication, which circumvents the need to set up the compute environment on CI, manage potential access keys, hardware requirements, etc.

* `CONTRIBUTING.qmd` / `FAQ.qmd` / `SETUP.qmd`: These are all pages on the website. You can see how they render on the website.

* `index.ipynb`: This is the publication.

* `demo.ipynb`: Originally, the contents of this notebook were placed in `index.ipynb`. However, after Ryan L playtested, we decided that it would be nice for the scientist to reference this demo during the development of their publication. For this reason, it's included in the navbar, and instructions are in the `README.md` that describe how it should be removed prior to publication.

* `Makefile`: `make preview` lets scientists view their publication as a live preview. `make execute` renders the publication (with cell execution) and updates `_freeze`. `make bump-version` manages a janky versioning system (see `CONTRIBUTING.qmd` for a blurb about it, and `_bump_version.py` for implementation).

* Everything else is styling. I'm sure many improvements to the code quality of `assets/` could be made. Since I'm not a front-end developer, it would be great to make any improvements to this section of the codebase through collaboration.

keithchev commented on _variables.yml:
```diff
@@ -0,0 +1,5 @@
+  org: "Arcadia-Science"
+  repo: "notebook-pub-template"
+  title: "This is the title for the notebook pub template"
+  google_analytics: "G-FMENQL377V"
```
Is the pub team is using the same GA ID for all of the pubs? (if not, we'd need to remove this GA ID)

ekiefl replied:
I guess you're right. This GA ID is for the notebook pub, but if people fail to update this with their own GA ID, we'll be concatenating analytics haphazardly.

keithchev commented on assets/author-reveal.html:
Was this file generated automatically, or was it written manually?

ekiefl replied:
This and everything in `assets/` is manually written.  Tried to clarify this and other info in developer-docs/TEMPLATE_ARCHITECTURE.md

keithchev commented on authors.yml:
```diff
@@ -0,0 +1,31 @@
+# - visualization
+authors:
+  - name: First M. Last
+    orcid: 0000-0002-6473-0921
```
Should probably use blank orcids here
```suggestion
    orcid: 0000-0000-0000-0000
```

keithchev commented on authors.yml:
```diff
@@ -0,0 +1,31 @@
+      - methodology
+      - software
+  - name: Contributor Name
+    orcid: 0000-0001-2345-6789
```
```suggestion
    orcid: 0000-0000-0000-0000
```

keithchev commented on pyproject.toml:
```diff
@@ -0,0 +1,73 @@
+"index*.ipynb" = [
+    # Markdown paragraphs are converted to single line comments that extend past line limit. Ignore.
+    "E501",
+    # Cells often end with useless expressions. Ignore.
```
```suggestion
    # Ignore useless expressions, because notebook cells often end with them.
```

keithchev commented on pyproject.toml:
```diff
@@ -0,0 +1,73 @@
+# ignore star and unused imports
+"__init__.py" = ["F403", "F405"]
+"index*.ipynb" = [
+    # Markdown paragraphs are converted to single line comments that extend past line limit. Ignore.
```
```suggestion
    # Ignore the line length limit because markdown paragraphs are converted to single line comments that may extend past the limit.
```

keithchev commented on ref.bib:
This file contains specific references that don't seem connected to the demo notebook. Are they included as examples?

ekiefl replied:
This is an outdated comment, but all references are referenced in the demo (now).

keithchev commented on assets/arcadia.csl:
Where did this file come from? Would be good to document this somewhere.

ekiefl replied:
Provenance added to TEMPLATE_ARCHITECTURE.md

keithchev commented on README.md:
````diff
@@ -0,0 +1,153 @@
+    pip install -e .
+    ```
+
+    And finally, install the [pre-commit](https://pre-commit.com/) hooks:
+
+    ```bash
+    pre-commit install
+    ```
````
Might be worth pointing out that this is optional, as they can be troublesome and not everyone likes/uses them.

keithchev commented on README.md:
```diff
@@ -0,0 +1,153 @@
+
+    Populate `README_TEMPLATE.md` and then rename it to `README.md`.
+
+    **NOTE**: The content you're reading now is the current `README.md`, which is to be replaced with `README_TEMPLATE.md`.
```
Does this mean the user should delete this document? That doesn't seem ideal, because it contains a lot of useful instructions.

ekiefl replied:
README_TEMPLATE content has been merged into README, and any important info in README is outsourced.

keithchev commented on README.md:
```diff
@@ -0,0 +1,153 @@
+
+1. Make the repository public
+
+    In order for this pub to be open and reproducible, make the [repo public](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/setting-repository-visibility).
```
Could we add a link to our standards for public repos here?

https://github.com/Arcadia-Science/arcadia-software-handbook/blob/main/guides-and-standards/standards--public-repos.md

keithchev commented on README.md:
```diff
@@ -0,0 +1,153 @@
+
+    Once happy, commit `index.ipynb` and all files in the `_freeze/` directory.
+
+    Now, create a pull request to merge your branch into `main`. Once your PR is approved, merge into `main`.
```
I got a bit confused here. This kinda makes it sound like merging into `main` is a final step before publishing, but I would assume most folks would be merging into `main` all along, as they write the pub, before they get to this step and make the repo public. 

Also, if the PR is just merging a snapshot of the notebook, it probably doesn't need to reviewed or approved, right?

ekiefl replied:
Will be revamped during git tag versioning refactor

keithchev commented on README.md:
```diff
@@ -0,0 +1,153 @@
+
+1. Host the publication
+
+    Publishing is automated through a GitHub Action that triggers when a pull request is merged into the `publish` branch. Thus, all that's required for your publication to go live is to merge into `publish`. By convention, we only merge changes from the `main` branch into `publish`. This ensures collaborators can merge their completed work into `main`, where others can see and build upon it, whi ...
```
What is meant by "private" here? If the repo itself is public, then nothing will be private (just published or unpublished).

ekiefl replied:
Will be revamped during git tag versioning refactor

keithchev commented on README.md:
```diff
@@ -0,0 +1,153 @@
+
+    Publishing is automated through a GitHub Action that triggers when a pull request is merged into the `publish` branch. Thus, all that's required for your publication to go live is to merge into `publish`. By convention, we only merge changes from the `main` branch into `publish`. This ensures collaborators can merge their completed work into `main`, where others can see and build upon it, whi ...
+
+    When all your changes have been merged into `main` and you're ready for your publication to go live, open a pull request to merge `main` into `publish`. Once approved and merged, your publication will be live within minutes.
```
Who approves and merges this PR? Is it the user themselves or do they need approval from the pub team?

Also, what will the URL of the publication be? Might be nice to mention that here.

ekiefl replied:
Will be revamped during git tag versioning refactor

keithchev commented on SETUP.qmd:
```diff
@@ -0,0 +1,80 @@
+
+# Setup
+
+This document details how to create a local copy of this pub's codebase and setup your compute environment. This is a prerequisite for reproducing this publication, experimenting with the analysis, or contributing a revision.
```
This document also describes how to reproduce the pub (and indeed is the target of the "reproduce" link in the navbar), so it feels weird not to mention that here.
```suggestion
This document details how to create a local copy of this pub's codebase, setup your compute environment, and reproduce the pub itself. This will enable you to experiment with the analysis in the pub and, optionally, contribute revisions to it.
```

keithchev commented on README.md:
```diff
@@ -0,0 +1,153 @@
+    * Replace the variables in `authors.yml`.
+        - You'll update this once contributor roles have been assigned by Publishing Team, but for now, feel free to add yourself.
+
+1. Install Quarto
+
+    The publication is rendered with [Quarto](https://quarto.org/). If you don’t have it installed (check with quarto --version), you can install it [here](https://quarto.org/docs/get-started/).
+
+1. Setup the code environment
```
I think these steps should be moved into `README_TEMPLATE.md`, because they're not really about how to use the template, but instead about to use the repos created from the templates. I think we'd also want to make sure these sections are preserved when the user removes this README and replaces it with the README_TEMPLATE.

ekiefl replied:
README_TEMPLATE content has been merged into README, and any important info in README is outsourced.

keithchev commented on README.md:
```diff
@@ -0,0 +1,153 @@
+    Edit `index.ipynb` to create your pub. As you work, render it in a live preview with `make preview`.
+
+
+## How to publish
```
I think I would move this section about publishing into its own document, so that it doesn't get lost when the user replaces this file with the README_TEMPLATE.

ekiefl replied:
Done

keithchev commented on SETUP.qmd:
```diff
@@ -0,0 +1,80 @@
+
+The codebase is hosted on GitHub and can be found [here](https://github.com/{{< var pub.org >}}/{{< var pub.repo >}}).
+
+To obtain a local copy, you have the option of either *cloning* or *forking* the repository. If you aren't sure what's best, our suggestion is to clone unless you both (1) want to propose a revision for the publication and (2) are not an employee of Arcadia Science.
```
Nit: they can either clone the repo or fork it and then clone their fork.
```suggestion
To obtain a local copy of this repo, you can either *clone* it directly or fork it to your own GitHub account, then clone your fork. If you aren't sure what's best, our suggestion is to clone directly unless you both (1) want to propose a revision for the publication and (2) are not an employee of Arcadia Science.
```

keithchev reviewed (commented):
This looks like a lot of work, and the rendered result looks great!

I have two major comments:
- I think the current README is doing too many things: it describes how to use the template, how to set up a dev env, and how to publish the pub. I think the sections about env setup and publishing could be moved to separate documents, both for clarity and also because the README directs the user to replace itself with README_TEMPLATE, and and I think the information about env setup and publishing should be retained.
- This could be done in a separate later PR, but I think some dev docs about the template itself would be important if we maintain or continue to develop this template. It looks like there are a lot of interacting third-party dependencies, and in many cases, it was hard for me to tell which files were vendored in, which were generated automatically (and/or should not be edited directly), and which were written from scratch. For this reason, I didn't review the files in `assets/` or `_extensions/`. 

Other questions/comments:
- It seems like the build env is currently the same as the analysis env (i.e. the single `env.yml` file). I could imagine this could become troublesome, but it seems like it may be necessary because the build workflow must actually execute the notebook. In any case, it might be good to make a note of this constraint somewhere.
- When running the pub in preview mode, the cookie consent banner appears on every page, even after clicking "accept".
- Just a note that the duplication of some material from the README_TEMPLATE in SETUP.qmd will be annoying to maintain, but I'm not sure I see a way to avoid it.

ekiefl commented:
Thanks for the feedback. I'll start making my way through the comments over the coming weeks.

To quickly touch on your two major points of feedback. I completely agree. What do you think about moving all developer instructions to this repo's GitHub Wiki? We could have pages for setting up the environment, publishing, review process, and whatever else. Is this something Software has the bandwidth to collaborate with me on?

keithchev commented:
> What do you think about moving all developer instructions to this repo's GitHub Wiki? We could have pages for setting up the environment, publishing, review process, and whatever else. Is this something Software has the bandwidth to collaborate with me on?

I'm not sure this will work because I don't think the wiki will be inherited by repos created from the template. The line between material that's specific to the template and material that will remain relevant to the "child repos" feels a little burry so it might be easiest to keep everything markdown files in the repo, in some obvious place like `developer-docs/`.

wasimsandhu commented on .github/workflows/publish.yml:
It would be good to add a link to the [instructions for setting up a GitHub Action for Quarto](https://quarto.org/docs/publishing/github-pages.html#github-action) in a comment here.

ekiefl replied:
Great plug, thanks. Added.

wasimsandhu reviewed (changes requested):
This is really great work. I feel like you've established a very seamless flow for notebook-to-pub development. I tried to summarize my suggestions (mostly around organization) in a concise manner below:

1. I agree with Keith that the main README should only contain instructions for how to use the template. You could create a folder called `development_docs` with `environment_setup.md` and `publishing.md`, and link to these in the main README.

2. I also agree that some meta-docs about the template would be useful (specifically the usage of the extensions and assets). Maybe `extensions.md` in `/_extensions` and `styles.md` in `/assets`.

3. I'm not sure `README_TEMPLATE.md` is needed. You could just instruct the user to delete the template usage instructions and fill out the template at the bottom of the same `README.md`.

4. I think the Quarto Markdown files that are used to render the pages should be moved to a `/pages` directory.

5. Similarly, I think the YML files (excluding `env.yml`) could be moved to a `/config` directory.

6. Finally, the `demo.ipynb` could be moved to an `/examples` directory (which we could populate with more demos in the future!).

Again, nice work!

ekiefl commented on Makefile:
```diff
@@ -0,0 +1,35 @@
+
+.PHONY: execute
+execute:
+	# Project-wide render without executing code cells. Instead, rely on
+	# pre-computed results present in _freeze/
+	quarto render
+
+	# Now, render `index.ipynb`, with code execution. This will populate
+	# _freeze/index/ with pre-computed results.
+	quarto render index.ipynb --execute
```
Discussed offline

ekiefl commented on CONTRIBUTING.qmd:
````diff
@@ -0,0 +1,56 @@
+
+If you haven't already, follow our [setup guide](SETUP.qmd) to create a local copy of the code and compute environment.
+
+## Snapshot current version
+
+Before making any changes, we need to preserve the current state of the publication by creating a versioned snapshot. This ensures we maintain an archive of all previous versions alongside the latest one. To create this snapshot, run:
+
+```bash
+make bump-version
+```
````
This will be fixed with our git tag-based versioning system.

ekiefl commented on CONTRIBUTING.qmd:
```diff
@@ -0,0 +1,56 @@
+
+To publish your revisions, we need you to open a pull request. And in order for us to merge your pull request, here's what we'll need from you in addition to your content changes.
+
+Begin with a clean branch (no uncommitted changes). Then run the notebook from the command line:
```
Resolved with git tag versioning system

ekiefl commented on ref.bib:
Because this is both a template and a demo, the `ref.bib` is populated with references used in the demo. I don't see a way around this.

ekiefl commented on env.yml:
```diff
@@ -0,0 +1,20 @@
+  - pre_commit>=4.0.1
+  - ruff>=0.6.9
+
+  # Notebook publication infrastructure
+  - jupyterlab>=4.2.5
+  - matplotlib>=3.9.2
+  - pyyaml>=6.0.2
+  - pandas>=2.2.3
```
I'll try and better separate true infrastructure from requirements of the demo, and pin the requirements of the demo.

ekiefl commented on CONTRIBUTING.qmd:
````diff
@@ -0,0 +1,56 @@
+make bump-version
+```
+
+Commit the changes and untracked files (`_freeze/*`, `_quarto.yml`, `index_*.ipynb`) introduced by this version snapshot before proceeding with your changes.
````
We can rework this section once new versioning in place

ekiefl commented on CONTRIBUTING.qmd:
````diff
@@ -0,0 +1,56 @@
+make execute
+```
+
+This command will update `index.ipynb` with the latest execution results. Importantly, it may generate runtime artifacts in the `_freeze/` directory.
````
Agreed. It's just important that they commit the _freeze results, which is described a few steps down.

ekiefl commented on FAQ.qmd:
```diff
@@ -0,0 +1,25 @@
+
+## How can I reproduce this?
+
+Reproducing this publication is easy. Follow this [setup guide](SETUP.qmd) to get started.
```
Edited to "Reproducing this publication is as easy as issuing a few commands from the command line". People can interpret that as easy or hard depending on their familiarity with the command line.

ekiefl commented on CONTRIBUTING.qmd:
```diff
@@ -0,0 +1,56 @@
+
+# Contributing
+
+We welcome improvements to this publication! If you'd like to improve or extend the publication, please submit a pull request. We'll collaborate with you to incorporate your revisions.
```
Added

ekiefl commented on SETUP.qmd:
````diff
@@ -0,0 +1,80 @@
+
+The best way to ensure you've correctly set up your code and compute environment is to reproduce this work. Fortunately, the analysis, and therefore the publication itself, can be reproduced with the following command:
+
+```bash
+make execute
````
Added disclaimer

ekiefl commented on _quarto.yml:
```diff
@@ -0,0 +1,84 @@
+    right:
+    - text: FAQ
+      href: FAQ.qmd
+    - text: Reproduce
```
Added

ekiefl commented on FAQ.qmd:
```diff
@@ -0,0 +1,25 @@
+
+## What is this?
+
+This *notebook publication* uses a format we're experimenting with that treats a scientist's working computational analysis as the publication itself, dissolving the separation that exists between code and publication. Our hypothesis is that this lets us publish faster, promote early-stage work, and increase reproducibility. For details, see [our commentary](https://doi.org/10.57844/arcadia-ca21- ...
```
It's correct, but correctly dead.

ekiefl commented on FAQ.qmd:
```diff
@@ -0,0 +1,25 @@
+
+## Where is the data/code?
+
+All the code and data for this publication and its analysis are hosted on GitHub at [this URL](https://github.com/{{< var pub.org >}}/{{< var pub.repo >}}).
```
The data will be hosted or pointed to in the repo, so I guess the sentence as it's written right now is incorrect. Edited to

> All the code for this publication and its analysis are hosted on GitHub at [this URL](https://github.com/{{< var pub.org >}}/{{< var pub.repo >}}). Any associated data is either hosted or linked to from this repository.

FYI: https://github.com/Arcadia-Science/arcadia-software-handbook/blob/main/guides-and-standards/standards--source-control.md#data-management

ekiefl commented on _extensions/mcanouil/iconify/_extension.yml:
It's kind of refreshing how simple it is.

wasimsandhu replied:
Until there's a couple 10,000 line extensions that need to be installed to implement a feature. 😬

ekiefl commented:
Thanks for your reviews. I've addressed everything, but two somewhat major things things remain.

First: shoddy versioning. @keithchev and I have a plan for cleaning up the versioning system. It will rely on git tags. @keithchev is going to work on this. There are some FIXME's to indicate documentation that will need to be reworked once those changes land.

Second: messy directory structure. @wasimsandhu, I agree the directory structure is flat and messy. There is a lot to be gained by a more hierarchical structure, particularly not overwhelming the scientist. That said, I had previously tried to address this but it didn't go over well. From what I remember, `index.ipynb` has to be in the same place as `_quarto.yml`, and all pages that load any of the custom HTML found in `assets/` need to be in the same directory because of hardcoded asset paths in those HTML files. For example, `logo-animation.html` references the path `./assets/logo_movie.mp4`, but because all pages load `logo-animation.html`, all these pages have to live in the same directory. If you wanted to take a stab at improving this, it would be appreciated.

wasimsandhu commented:
@ekiefl Here's what I did:

1. Move `.qmd` files to `/pages`.
2. Change the following in `_quarto.yml`:

```yml
website:
  favicon: "assets/logo_white.png"
  navbar:
  ...
    right:
    - text: FAQ
      href: pages/FAQ.qmd
    - text: Reproduce this pub
      href: pages/SETUP.qmd
    - text: Contribute
      href: pages/CONTRIBUTING.qmd
  ...
  page-footer:
    background: '#484B50'
    border: false
    left: '[![](../assets/logo_white.png){fig-alt="Arcadia-Science" width=65px}](https://www.arcadiascience.com/)
```

3. In the HTML files, set the paths like so:

```js
source.src = "../../assets/logo_movie.mp4";
```

This seems to fix the logo animation, and the links in the preview site appear to be working too.

wasimsandhu commented on assets/logo-animation.html:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    // Add hover event listener
+    brandLink.addEventListener("mouseenter", handleVideoPlayback);
```
You can remove this event listener if you don't want the animation to replay every time the user hovers over the logo. :)

wasimsandhu commented on CONTRIBUTING.qmd:
Sigh, I know... It took me some trial and error to figure out the correct relative paths. I left a [comment](https://github.com/Arcadia-Science/notebook-pub-template/pull/7#issuecomment-2702357410) in the main PR thread about how to achieve this.

wasimsandhu commented:
> From what I remember, index.ipynb has to be in the same place as _quarto.yml...

I was able to move `demo.ipynb` to an `/examples` directory and access it from the preview server once I made the following change in `_quarto.yml`:

```yml
- text: Demo
  href: /examples/demo.html
```

Note the `.html` instead of `.ipynb` extension for the filename.

wasimsandhu reviewed (commented):
The `developer-docs` are so satisfying to read. Great job with the reorg.

Left some comments about how to move the `.qmd` files into `/pages`, as well as `demo.ipynb` to `/examples`. I'll hold off on approving for now since it seems like there are more changes necessary.

ekiefl commented:
Thank you @wasimsandhu, your patch worked.

Besides a new versioning system, I think I've addressed all concerns. Assuming so, should we merge this into `empty-repo`, @keithchev, and treat that as a long-term dev branch for you to eventually merge the versioning rehaul into?

keithchev commented:
> Besides a new versioning system, I think I've addressed all concerns. Assuming so, should we merge this into empty-repo, @keithchev, and treat that as a long-term dev branch for you to eventually merge the versioning rehaul into?

yep, sounds good to me!

keithchev commented on developer-docs/TEMPLATE_ARCHITECTURE.md:
This is a delightful piece of documentation 😍

keithchev reviewed (approved):
This looks great! The new organization of the documentation and the directory structure look super clear and thorough.

wasimsandhu reviewed (approved):
🚀
